components/object_store/src/obkv/mod.rs

```diff
@@ -0,0 +1,914 @@
+    shard_num: usize,
+    table_names: Vec<String>,
+}
+
+impl ShardManager {
+    fn try_new<T: TableKv>(client: Arc<T>, shard_num: usize) -> std::result::Result<Self, Error> {
+        let mut table_names = Vec::with_capacity(shard_num);
+
+        for shard_id in 0..shard_num {
+            let table_name = format!("object_store_{shard_id}");
+            create_table_if_not_exists(&client, &table_name)?;
+            table_names.push(table_name);
+        }
+
+        Ok(Self {
+            shard_num,
+            table_names,
+        })
+    }
+
+    #[inline]
+    pub fn pick_shard_table(&self, path: &Path) -> &str {
+        let mut hasher = XxHash64::with_seed(HASH_SEED);
+        path.as_ref().as_bytes().hash(&mut hasher);
+        let hash = hasher.finish();
+        let index = hash % (self.table_names.len() as u64);
+        &self.table_names[index as usize]
+    }
+}
+
+impl std::fmt::Display for ShardManager {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ObjectStore ObkvShardManager({})", self.shard_num)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct ObkvObjectStore<T> {
+    /// The manager to manage shard table in obkv
+    shard_manager: ShardManager,
+    /// The manager to manage object store meta, which persist in obkv
+    meta_manager: Arc<MetaManager<T>>,
+    client: Arc<T>,
+    current_upload_id: AtomicI64,
+    /// The size of one object part persited in obkv
+    /// It may cause problem to save huge data in one obkv value, so we
+    /// need to split data into small parts.
+    part_size: usize,
+    /// The max size of bytes, default is 1GB
+    max_object_size: usize,
+    /// Maximum number of upload tasks to run concurrently
+    max_upload_concurrency: usize,
+}
+
+impl<T: TableKv> std::fmt::Display for ObkvObjectStore<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "ObkvObjectStore({:?},{:?})",
+            self.client, self.shard_manager
+        )?;
+        Ok(())
+    }
+}
+
+impl<T: TableKv> ObkvObjectStore<T> {
+    pub fn try_new(
+        shard_num: usize,
+        part_size: usize,
+        client: Arc<T>,
+        max_object_size: usize,
+        max_upload_concurrency: usize,
+    ) -> Result<Self> {
+        let shard_manager = ShardManager::try_new(client.clone(), shard_num).map_err(|source| {
+            StoreError::Generic {
+                store: OBKV,
+                source: Box::new(source),
+            }
+        })?;
+        let meta_manager: MetaManager<T> =
+            MetaManager::try_new(client.clone()).map_err(|source| StoreError::Generic {
+                store: OBKV,
+                source: Box::new(source),
+            })?;
+        let upload_id = AtomicI64::new(0);
+        Ok(Self {
+            shard_manager,
+            meta_manager: Arc::new(meta_manager),
+            client,
+            current_upload_id: upload_id,
```

Comment:
I guess AtomicU64 is better:
```suggestion
        Ok(Self {
            shard_manager,
            meta_manager: Arc::new(meta_manager),
            client,
            current_upload_id: AtomicU64::new(0),
```